### PR TITLE
Fix command signs

### LINKED ISSUE #6138: clickEvent:action run_command tellraw gives permission error for players
### Expected behavior

the tellraw, when clicked the sign as player

### Observed/Actual behavior

I'm sorry, but you do not have permission to perform this command. 

### Steps/models to reproduce

`/give playername oak_sign{BlockEntityTag:{ Text1:"{\"text\":\"---------------\", \"clickEvent\":{\"action\":\"run_command\",\"value\":\"/tellraw @p [{\\\"text\\\":\\\"test\\\",\\\"color\\\":\\\"yellow\\\",\\\"bold\\\":true},{\\\"text\\\":\\\"\\\\ntest\\\",\\\"color\\\":\\\"white\\\",\\\"bold\\\":false}]\"}}", Text2:'{"text":"test","color":"yellow"}', Text3:'{"text":"test","color":"white"}', Text4:'{"text":"---------------"}'}, display:{Name:'{"text":"Custom Sign"}'}, id:"Sign"}`

### Plugin list

no plugins, testserver

### Paper version

paper-1.17.1-91

### Agreements

- [X] I am running the latest version of Paper available from https://papermc.io/downloads.
- [X] I have searched for and ensured there isn't already an open issue regarding this.
- [X] My version of Minecraft is supported by Paper.

### Other

https://github.com/PaperMC/Paper/commit/648f6078d77876217b90dcea644a4e5c68c2e132

### LINKED ISSUE #6153: sign clickEvent run_command excuted where player is standing
### Expected behavior

the sign disappears when player clicks the sign that has clickevent {"action":"run_command","value":"setblock ~ ~ ~ air"}

### Observed/Actual behavior

the sign does not disappear except when i stand on that sign

### Steps/models to reproduce

1. /give @s minecraft:birch_sign{BlockEntityTag:{Text2:'{"text":"click me", "clickEvent":{"action":"run_command","value":"setblock ~ ~ ~ air"}}'}} 1
2. place the sign
3. click the sign

### Plugin list

no plugins

### Paper version

paper 1.17.1-99

### Agreements

- [X] I am running the latest version of Paper available from https://papermc.io/downloads.
- [X] I have searched for and ensured there isn't already an open issue regarding this.
- [X] My version of Minecraft is supported by Paper.

### Other

https://github.com/PaperMC/Paper/commit/648f6078d77876217b90dcea644a4e5c68c2e132
https://github.com/PaperMC/Paper/issues/6138

### PR DESCRIPTION
**Pull request by Machine-Maker**
Fixes #6138

This should address all of @mdcfe points of issue with how they were handled before, but keeping the vanilla permissions behavior.

### CODE REVIEW

File: build-data/paper.at [hunk 1]
```diff
@@ -231,3 +231,6 @@ public net.minecraft.world.level.entity.PersistentEntitySectionManager sectionSt
 
 # Optimize light engine
 public-f net.minecraft.world.level.chunk.DataLayer
+
+# Change handling of sign run_command click events
+public net.minecraft.server.network.ServerGamePacketListenerImpl LOGGER
```
Comment (Proximyst) [inline build-data/paper.at]: Create a new logger instead

File: patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+--- a/src/main/java/net/minecraft/world/level/block/entity/SignBlockEntity.java
++++ b/src/main/java/net/minecraft/world/level/block/entity/SignBlockEntity.java
+@@ -227,7 +227,15 @@ public class SignBlockEntity extends BlockEntity implements CommandSource { // C
+             ClickEvent chatclickable = chatmodifier.getClickEvent();
+ 
+             if (chatclickable != null && chatclickable.getAction() == ClickEvent.Action.RUN_COMMAND) {
+-                player.getServer().getCommands().performCommand(this.createCommandSourceStack(player), chatclickable.getValue());
++                // Paper start
++                net.minecraft.server.network.ServerGamePacketListenerImpl.LOGGER.info(player.getScoreboardName() + " issued server command: /" + chatclickable.getValue());
++                org.bukkit.event.player.PlayerCommandPreprocessEvent event = new org.bukkit.event.player.PlayerCommandPreprocessEvent(player.getBukkitEntity(), "/" + chatclickable.getValue(), new org.bukkit.craftbukkit.util.LazyPlayerSet(player.getServer()));
```
Comment (Proximyst) [inline patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch]: Perhaps a subclass would be preferred, such that it is clear that in this case, a certain sign is being used as a context and that it _will_ bypass permissions?

File: patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
++                final String command = event.getMessage().substring(1);
++                if (!event.callEvent()) {
++                    return false;
++                }
++                player.getServer().getCommands().performCommand(this.createCommandSourceStack(((org.bukkit.craftbukkit.entity.CraftPlayer) event.getPlayer()).getHandle()), command, command, false);
```
Comment (Proximyst) [inline patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch]: Doesn't this keep the problem of commands from plugins not working, and any failures being silent?

File: patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+index 9b5d11ece006d7aa893360a84ba652c666517ac1..81739da3afd5979ecf745dce39d547c8dcc07250 100644
+--- a/src/main/java/net/minecraft/world/level/block/entity/SignBlockEntity.java
++++ b/src/main/java/net/minecraft/world/level/block/entity/SignBlockEntity.java
+@@ -227,7 +227,15 @@ public class SignBlockEntity extends BlockEntity implements CommandSource { // C
+             ClickEvent chatclickable = chatmodifier.getClickEvent();
+ 
+             if (chatclickable != null && chatclickable.getAction() == ClickEvent.Action.RUN_COMMAND) {
+-                player.getServer().getCommands().performCommand(this.createCommandSourceStack(player), chatclickable.getValue());
++                // Paper start
++                net.minecraft.server.network.ServerGamePacketListenerImpl.LOGGER.info(player.getScoreboardName() + " issued server command: /" + chatclickable.getValue());
```
Comment (jpenilla) [inline patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch]: This should respect the config option same as the log message for normal commands.

File: patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
++package io.papermc.paper.commands;
++
++import net.minecraft.commands.CommandSource;
++import net.minecraft.commands.CommandSourceStack;
++import net.minecraft.network.chat.Component;
++import org.bukkit.command.CommandSender;
++
++import java.util.UUID;
++
++public class DelegatingCommandSource implements CommandSource {
```
Comment (Proximyst) [inline patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch:42]: `DelegatingBukkitToMinecraftCommandSource`

File: patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch (hunk 2 above)
Comment (jpenilla) [inline patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch]: Don't we only want to add a slash if it's not already starting with one?

**Review by Proximyst — CHANGES_REQUESTED** [review]

File: patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+         }
+ 
+@@ -263,8 +271,21 @@ public class SignBlockEntity extends BlockEntity implements CommandSource { // C
+         String s = player == null ? "Sign" : player.getName().getString();
+         Object object = player == null ? new TextComponent("Sign") : player.getDisplayName();
+ 
++        // Paper start - send messages back to the player
++        CommandSource commandSource = new io.papermc.paper.commands.DelegatingCommandSource(this) {
++            @Override
++            public void sendMessage(net.minecraft.network.chat.Component message, java.util.UUID sender) {
```
Comment (jpenilla) [inline patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch:113]: Why are these overrides needed here if they are overridden in DelegatingCommandSource

**Review by jpenilla — DISMISSED** [review]

File: patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch (hunk 3 above)
Comment (Machine-Maker) [reply patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch]: Nope, commands from plugins work just fine. (I tested it) and failures are sent to the CommandSource (which is why I created a new one and overrode the sendMessage to send it to the player). Tested that too with an unknown command.

File: patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch (hunk 6 above)
Comment (Machine-Maker) [reply patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch:113]: Because it's redirecting the message to be send to the player instead of the delegate CommandSource. In this case, the SignBlockEntity.

File: patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch (hunk 2 above)
Comment (jpenilla) [reply patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch]: Might be out of scope for this pr, but exposing that it was a sign command to plugins could be useful

File: patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch (hunk 4 above)
Comment (Proximyst) [reply patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch]: Is there not some game rule for stuff like this too, or is that only for certain commands and command blocks?

File: patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch (hunk 3 above)
Comment (Proximyst) [reply patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch]: Does this also work for WorldEdit-y commands (i.e. ones with double slash)?

Comment (Machine-Maker) [reply patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch]: Yep, tested that too.
just created these commands to test them in a test plugin
```yaml
commands:
  test:
    description: Test Command
  /slashtest:
    description: Slash Test Command
```

File: patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch (hunk 4 above)
Comment (jpenilla) [reply patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch]: commands are not logged at all by vanilla, it's an option in the spigot config and spigot added log message

File: patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch (hunk 2 above)
Comment (Machine-Maker) [reply patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch]: Well, I was thinking there should be a CommandSenderCommandPreprocessEvent for handling commands from CommandSenders that *arent* players. And this would fall into that category possibly.

File: patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch (hunk 4 above)
Comment (jpenilla) [reply patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch]: and yes, the gamerule is for when `sendSuccess(Component, boolean)` is called with true for the broadcastToOps bool param, which is mostly only vanilla commands, and for command feedback, not logging the command which was ran

File: patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch (hunk 2 above)
Comment (jpenilla) [reply patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch]: For signs we'd probably want a subclass of the Player event, as it is a player executing the command after all. A generic `CommandPreprocessEvent` might be useful for other spots where there's a missing event, but in this case I don't see it being any better

File: patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch [hunk 7]
```diff
@@ -0,0 +1,103 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Jake Potrebic <jake.m.potrebic@gmail.com>
+Date: Fri, 9 Jul 2021 13:50:48 -0700
+Subject: [PATCH] Fix commands from signs not firing command events
+
+This patch changes sign command logic so that `run_command` click events:
+  - are logged to the console
+  - fire PlayerCommandPreprocessEvent
+  - work with double-slash commands like `//wand`
+  - sends failure messages to the player who clicked the sign
```
Comment (Phoenix616) [inline patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch:10]: Why is this behaviour change necessary? Doesn't this result in message spam with existing setups which relied on that not being the case?

Comment (Machine-Maker) [reply patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch:10]: Relied on the command throwing an exception?

Comment (Phoenix616) [reply patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch:10]: Is this really only for syntax errors/unknown commands? I was under the impression this would also catch cases where e.g. no entity would match a selector or a block could not be placed (a case Vanilla commands return "`false`")

File: patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch (hunk 2 above)
Comment (mdcfe) [inline patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch]: Just logging for future reference: in vanilla, signs on commands *may or may not* have a `/` at the start (ie both `tellraw @p ...` and `/tellraw @p ...` are valid). However, `PlayerCommandPreprocessEvent` always has a leading `/` when fired by `ServerGamePacketListenerImpl#handleCommand`, so this needs to check for the presence of a leading slash and prepend one if missing.

(As for double slash commands from WorldEdit like `//wand`, these never worked prior to #6109, but it makes sense to require two leading slashes to avoid ambiguity.)

**Review by mdcfe — CHANGES_REQUESTED** [review]

File: patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch (hunk 7 above)
Comment (Machine-Maker) [reply patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch:10]: Yeah, it looks like it catches any exception. So I'll make this a config that defaults to the failure messages being hidden.

File: patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch (hunk 2 above)
Comment (Machine-Maker) [reply patches/server/0723-Fix-commands-from-signs-not-firing-command-events.patch]: Ok, I added a PlayerSignCommandPreprocessEvent

File: patches/api/0323-Add-PlayerSignCommandPreprocessEvent.patch [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
++
++    private final Sign sign;
++
++    public PlayerSignCommandPreprocessEvent(@NotNull Player player, @NotNull String message, @NotNull Set<Player> recipients, @NotNull Sign sign) {
++        super(player, message, recipients);
++        this.sign = sign;
++    }
++
++    /**
++     * Returns the sign that the command originated from.
```
Comment (mbax) [inline patches/api/0323-Add-PlayerSignCommandPreprocessEvent.patch]: "Gets the sign from which the command originated."

**Review by mbax — CHANGES_REQUESTED** [review]
<3

**Review by Proximyst — APPROVED** [review]
LGTM!

**Review by mdcfe — APPROVED** [review]
<details>
<summary><b>This is inaccurate, see <a href="https://github.com/PaperMC/Paper/pull/6139#pullrequestreview-703606704">this</a></b></summary>

Tested with [EssentialsX 2.19.0-dev+197-5e17894](https://ci.ender.zone/job/EssentialsX/1235/) and [WorldEdit 7.2.6 Beta 1](https://dev.bukkit.org/projects/worldedit/files/3379705).

EssentialsX command cooldowns work as expected (listens to and cancels `PlayerCommandPreprocessEvent`) - eg `minecraft:tell` and `/minecraft:tell` both trigger the cooldown `'minecraft:tell': 10`.

## Command execution behaviour comparison

EssentialsX kit with below signs: https://paste.lucko.me/WJWEeW5YoS

| `run_command` value           | Before patch (build 85) | #6109 patch             | This patch              |
|-------------------------------|-------------------------|-------------------------|-------------------------|
| `"minecraft:tell @p hello"`   | Runs command            | Runs command            | Runs command            |
| `"/minecraft:tell @p hello"`  | Runs command            | Runs command            | Runs command            |
| `"//minecraft:tell @p hello"` | Silently fails          | Shows "Unknown command" | Shows "Unknown command" |
| `"suicide"`                   | Runs command            | Runs command            | Runs command            |
| `"/suicide"`                  | Shows "Unknown command" | Runs command            | Runs command            |
| `"//suicide"`                 | Silently fails          | Shows "Unknown command" | Shows "Unknown command" |
| `"wand"`                      | Silently fails          | Shows "Unknown command" | Shows "Unknown command" |
| `"/wand"`                     | Silently fails          | Shows "Unknown command" | Shows "Unknown command" |
| `"//wand"`                    | Shows "Unknown command" | Runs command            | Runs command            |
| `"///wand"`                   | Silently fails          | Shows "Unknown command" | Shows "Unknown command" |

TL;DR test cases work same as #6109 and do not regress on vanilla behaviour aside from improved command feedback.

</details>

**Review by MiniDigger — APPROVED** [review]

**Review by mdcfe — APPROVED** [review]
Re-reviewing as previous results were incorrect.

## Command execution behaviour comparison

Tested with [EssentialsX 2.19.0-dev+197-5e17894](https://ci.ender.zone/job/EssentialsX/1235/) and [WorldEdit 7.2.6 Beta 1](https://dev.bukkit.org/projects/worldedit/files/3379705).

EssentialsX command cooldowns work as expected (listens to and cancels `PlayerCommandPreprocessEvent`) - eg `minecraft:tell` and `/minecraft:tell` both trigger the cooldown `'minecraft:tell': 10`.

<details>
<summary>Tests</summary>

### EssentialsX kit

```yaml
kits:
  test:
    delay: 0
    items:
      - /minecraft:give {USERNAME} minecraft:oak_sign{BlockEntityTag:{Text1:'{"text":"minecraft:tell @p hello","clickEvent":{"action":"run_command","value":"minecraft:tell @p hello"}}'}}
      - /minecraft:give {USERNAME} minecraft:oak_sign{BlockEntityTag:{Text1:'{"text":"/minecraft:tell @p hello","clickEvent":{"action":"run_command","value":"/minecraft:tell @p hello"}}'}}
      - /minecraft:give {USERNAME} minecraft:oak_sign{BlockEntityTag:{Text1:'{"text":"//minecraft:tell @p hello","clickEvent":{"action":"run_command","value":"//minecraft:tell @p hello"}}'}}
      - /minecraft:give {USERNAME} minecraft:oak_sign{BlockEntityTag:{Text1:'{"text":"suicide","clickEvent":{"action":"run_command","value":"suicide"}}'}}
      - /minecraft:give {USERNAME} minecraft:oak_sign{BlockEntityTag:{Text1:'{"text":"/suicide","clickEvent":{"action":"run_command","value":"/suicide"}}'}}
      - /minecraft:give {USERNAME} minecraft:oak_sign{BlockEntityTag:{Text1:'{"text":"//suicide","clickEvent":{"action":"run_command","value":"//suicide"}}'}}
      - /minecraft:give {USERNAME} minecraft:oak_sign{BlockEntityTag:{Text1:'{"text":"sel","clickEvent":{"action":"run_command","value":"sel"}}'}}
      - /minecraft:give {USERNAME} minecraft:oak_sign{BlockEntityTag:{Text1:'{"text":"/sel","clickEvent":{"action":"run_command","value":"/sel"}}'}}
      - /minecraft:give {USERNAME} minecraft:oak_sign{BlockEntityTag:{Text1:'{"text":"//sel","clickEvent":{"action":"run_command","value":"//sel"}}'}}
      - /minecraft:give {USERNAME} minecraft:oak_sign{BlockEntityTag:{Text1:'{"text":"///sel","clickEvent":{"action":"run_command","value":"///sel"}}'}}
```

### Results

| `run_command` value           | Before patch (build 85)    | #6109 patch                | This patch (feedback off) | This patch (feedback on)                      |
|-------------------------------|----------------------------|----------------------------|---------------------------|-----------------------------------------------|
| `"minecraft:tell @p hello"`   | Runs command elevated      | Runs command as player     | Runs command elevated     | Runs command elevated                         |
| `"/minecraft:tell @p hello"`  | Runs command elevated      | Runs command as player     | Runs command elevated     | Runs command elevated                         |
| `"//minecraft:tell @p hello"` | Silently fails             | Shows CB "Unknown command" | Silently fails            | Shows vanilla "Unknown or incomplete command" |
| `"suicide"`                   | Runs command as player     | Runs command as player     | Runs command as player    | Runs command as player                        |
| `"/suicide"`                  | Shows CB "Unknown command" | Runs command as player     | Runs command as player    | Runs command as player                        |
| `"//suicide"`                 | Silently fails             | Shows CB "Unknown command" | Silently fails            | Shows vanilla "Unknown or incomplete command" |
| `"sel"`                       | Silently fails             | Shows CB "Unknown command" | Silently fails            | Shows vanilla "Unknown or incomplete command" |
| `"/sel"`                      | Silently fails             | Shows CB "Unknown command" | Silently fails            | Shows vanilla "Unknown or incomplete command" |
| `"//sel"`                     | Shows CB "Unknown command" | Runs command as player     | Runs command as player    | Runs command as player                        |
| `"///sel"`                    | Silently fails             | Shows CB "Unknown command" | Silently fails            | Shows vanilla "Unknown or incomplete command" |

**(TL;DR everything now works as I'd expect)**

</details>
